### PR TITLE
Refactor workflow to support PERs and ARs

### PR DESCRIPTION
**Pull request by Crell**
Let's finally define a process for "living specs", which I have here called PHP Evolving Recommendations (PERs).  This now involves a number of moving parts, which largely come down to refactoring bits of the PSR process out to separate pieces and reusing them more formally.

# Changelog

* Simplify voting somewhat by defining two base vote types: Approval (Yes/No) and Decision (multiple choice).
* Establish a process for Implicit Approval by the Core Committee to minimize pointless voting.
* Move the Working Group process out of the PSR workflow into its own section under Mission and Structure, complete with full lifecycle personnel management.
* Establish two "levels" of Working Group.
* Refactor the PSR workflow to take advantage of the above changes.
* Define a new workflow for PERs using the above independent bits.
* Explicitly list the types of artifact we produce.
* Defines the role of "Maintainer" for artifacts, including utility libraries plus PSR/PER output.
* Reorders the bylaws to put PERs up front by PSRs.

As a practical matter, basically nothing changes for PSRs from today.  The changes look larger because it's lots of refactoring out to base classes and composed objects rather than one big god object, but the unit tests for the PSR process should still all pass.  If, you know, that was a thing... :smile:

### CODE REVIEW

File: bylaws/009-per-workflow.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+A PSR may be Deprecated explicitly as part of the Acceptance Vote for another PSR. Alternatively, it may be marked Deprecated by a Deprecation Vote.
+
+## Abandoned
+
+A PER Working Group can be marked as Abandoned by Secretaries when it is without an Editor for 60 days. After a period of 6 months without significant activity in a Working Group, the Secretaries may also declare a PER Working Group to be Abandoned.
+
+Should in the future a new Editor wish to establish a new Working Group to continue development of the same PER, the candidate may petition the Core Committee to do so by asking for a new Entrance Vote.  Only the Editor and two additional individuals are required at this time.
+
+## PER Working Group Fast-track
```
Comment (Crell) [inline bylaws/009-per-workflow.md:83]: I don't like the name for this section.  Please suggest something better.

Comment (benedmunds) [reply bylaws/009-per-workflow.md:75]:
```suggestion
## Expedited PER Working Group
```

**Comment by benedmunds** [comment]
I'm not sure where to put it but it's probably worth a TLDR of the differences between PSRs and PERs without having to dive into the workflow to figure it out.

**Comment by Crell** [comment]
It might make sense to merge the workflows at some point, at least descriptively.  But that's more work so I didn't want to get into it unless we were certain we wanted to go that route. :smile: 

But the tldr is basically what's in the description now.  A PER is the same as a PSR except it gets a name instead of a number, and can have SemVer releases after it's approved.  3/4 of the new PER page is a copy-paste from the PSR page.

File: bylaws/009-per-workflow.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+The Editor is empowered to declare tagged versions of a PER at any time.  Tagged versions MUST follow Semantic Versioning conventions.  However, any change that would necessitate a new Major release MUST be put to an Acceptance Vote by the Core Committee.
+
+## Staffing Changes
+
+The Editor of an approved PER is empowered to add or remove members from the Working Group at any time, with notification to the Core Committee.
+
+Should the Editor of a PER Working Group resign or become inactive, the Core Committee may appoint a new Editor.  Preference will be given to the existing Working Group members at that time.
+
+## Abandoned
```
Comment (benedmunds) [inline bylaws/009-per-workflow.md:69]: Would an "Accepted" PER be marked as "Abandoned" if it's WG dissolves?  Or would it forever remain "Accepted"?

**Comment by benedmunds** [comment]
> But the tldr is basically what's in the description now
Ah yes, I didn't mean for our purposes here, I was thinking of a TLDR for readers.

File: bylaws/009-per-workflow.md (hunk 2 above)
Comment (Crell) [reply bylaws/009-per-workflow.md:69]: The PER would remain accepted, I presume. I suppose we probably do want to have a process to officially deprecate and abandon a PER's artifacts.

File: bylaws/009-per-workflow.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+* One Editor
+* One Core Committee member who will act as Sponsor
+* At least three other individuals. These may include Project Representatives, Secretaries or Core Committee members as well as members of the general community.
+
+The proposal is not required to be fully developed at this point, although that is permitted. At minimum, it must include a statement of the problem to be solved and basic information on the general approach to be taken. Further revision and expansion is expected during the Draft Phase.
+
+The Sponsor may then call for an Entrance Vote of the Core Committee to enquire whether the Core Committee is generally interested in maintaining a PER for the proposed subject, even if they disagree with the details of the proposal.
+
+If the vote passes, the proposal officially enters Draft stage. The proposal is given a unique descriptive name (such as "CodingStandards", "CacheUtil", etc.).
```
Comment (alekitto) [inline bylaws/009-per-workflow.md:21]: Naming convention: is there a reason for the descriptive names to be in camel case?
I thought that these names were meant for humans, allowing spaces to be used (ex `Coding Standards`). Defining the package name, kebab-case convention should be used.

File: bylaws/003-votes.md [hunk 4]
```diff
@@ -43,6 +43,10 @@ A Deprecation Vote is called by any Core Committee Member. Only Core Committee m
 
 An Abandonment Vote is called by any Secretary or Core Committee Member. Only Core Committee members may vote, either For (+1), Against (-1), or Abstain (+0). Quorum is 50%. A 2/3 majority is required for passage.
 
+## Decision Vote
+
+A Decision Vote is called by any Secretary or Core Committee Member.  Decision Votes are not yes-or-no questions but selecting between two or more mutually-exclusive courses of action.  A Decision Vote will be conducted using Single Transferable Vote with a single-winner.  Quorum is 50%.
```
Comment (vdelau) [inline bylaws/003-votes.md:48]: Who is eligible to vote? Quorum is 50% of what group? I assume CC-only for both but I think it should be specified like in the other paragraphs in this bylaw.

**Comment by vdelau** [comment]
Personally I believe there could be a clearer separation between the product (PSR, PER) and the process (WG). I can see a WG maintaining one or more products, and proposing new products where they identify a need or opportunity. Forming a new WG would probably require a vote to establish a WG charter and initial lead ('editor'). The charter should specify the purpose of the WG and the products it inherits (existing PSR's, PER's) and what products it should deliver (newly proposed PSR's and PER's). Introducing a new product also requires a vote and establish which WG is responsible for the product and who the editor/maintainer and sponsor are.

File: bylaws/009-per-workflow.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+Unless a proposal is moved to Draft stage again, it must remain in Review stage for a minimum of four weeks and until two independent viable trial implementations can be demonstrated. The responsibility for finding such trial implementations and presenting them to the Core Committee lies with the Working Group, and especially the Editor and Sponsor. As not all specifications are PHP interfaces where the definition of "implementation" is self-evident, the Sponsor should use good faith judgement to determine when that is the case. Any member of the Core Committee may object to a given trial implementation as irrelevant or insufficient with due cause.
+
+Once four weeks have passed and two viable trial implementations can be demonstrated, the Sponsor may call an Acceptance Vote. If the Acceptance Vote fails, the specification may remain in Review.
+
+## Accepted
+
+If the Acceptance Vote passes, then the proposal officially becomes an accepted PER.  It is considered the 1.0.0 release of that PER.
```
Comment (Jean85) [inline bylaws/009-per-workflow.md:51]: I would reword "considered" into something more tangible, along the lines of "this is the moment where the editor will effectively tag the 1.0".

File: bylaws/007-psr-naming-conventions.md [hunk 6]
```diff
@@ -5,17 +5,18 @@ Naming conventions for code released by PHP FIG
 2. Abstract classes MUST be prefixed by `Abstract`: e.g. `Psr\Foo\AbstractBar`.
 3. Traits MUST be suffixed by `Trait`: e.g. `Psr\Foo\BarTrait`.
 4. PSR-1, 4, and 12 MUST be followed.
-5. The vendor namespace MUST be `Psr`.
-6. There MUST be a package/second-level namespace in relation with the PSR that
+5. The vendor namespace for code released as part of a PSR MUST be `Psr`.
+6. The vendor namespace for code released as part of a PER MUST be `Fig`.
```
Comment (shadowhand) [inline bylaws/007-psr-naming-conventions.md:9]: Why the discrepancy here? If a PSR namespace is `Psr`, why wouldn't a PER namespace be `Per`?

File: bylaws/007-psr-naming-conventions.md [hunk 7]
```diff
@@ -5,17 +5,18 @@ Naming conventions for code released by PHP FIG
...
+7. There MUST be a package/second-level namespace in relation with the PSR or PER that
    covers the code.
-7. Composer package MUST be named `psr/<package>` e.g. `psr/log`. If they
+8. Composer packages for PSRs MUST be named `psr/<package>` e.g. `psr/log`. Composer packages for PERs MUST be named `fig/<package>`.  If they
```
Comment (shadowhand) [inline bylaws/007-psr-naming-conventions.md:12]: Same question here.

File: bylaws/007-psr-naming-conventions.md [hunk 8]
```diff
@@ -5,17 +5,18 @@ Naming conventions for code released by PHP FIG
...
    require an implementation as a virtual package it MUST be named
    `psr/<package>-implementation` and be required with a specific version like
    `1.0.0`. Implementors of that PSR can then provide
    `"psr/<package>-implementation": "1.0.0"` in their package to satisfy that
    requirement. Specification changes via further PSRs should only lead to a new
    tag of the `psr/<package>` package, and an equal version bump of the
    implementation being required.
-8. Special lightweight utility packages MAY be produced alongside PSRs and
+9. Special lightweight utility packages MAY be produced alongside PSRs and
```
Comment (shadowhand) [inline bylaws/007-psr-naming-conventions.md:20]:
```suggestion
9. Special lightweight utility packages MAY be produced alongside PSRs and PERs, and
```

File: bylaws/009-per-workflow.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
+# PER Workflow
+
+A PHP Evolving Recommendation (PER) is a formal definition of best practices, utilities, reference implementations, and similar tools.  They may evolve over time as the PHP ecosystem develops.
```
Comment (shadowhand) [inline bylaws/009-per-workflow.md:3]: I like the wording here, and I think it might be helpful to expand on what "ecosystem develops" means; new features added to the language, deprecations, etc. Maybe something as simple as:

> as the PHP ecosystem develops due to new PHP features and deprecations.

File: bylaws/009-per-workflow.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+For any proposed change to the PER, the Editor must notify the Core Committee via a post to the mailing list of an Intent to Merge a given change.
+
+If no member of the Core Committee objects within seven days, the proposed change is assumed accepted by default.  Any member of the Core Committee may request during that seven-day period that the change be put to a vote.  If any member does so, the change is immediately put to an Acceptance Vote to determine if it is accepted or not.
+
+The Editor is empowered to declare tagged versions of a PER at any time.  Tagged versions MUST follow Semantic Versioning conventions.  However, any change that would necessitate a new Major release MUST be put to an Acceptance Vote by the Core Committee.
```
Comment (shadowhand) [inline bylaws/009-per-workflow.md:61]: Along with my previous comment, I am not sure that "any time" is a good idea here. Perhaps this should be more defined, per SemVer:

1. Bug fixes MAY be tagged by the Editor at any time.
2. Minor versions MUST be declared to the Core Committee and have a 7 day holding period, during which any Core Committee member MAY call for an Acceptance Vote.
3. Major versions MUST be declared to the Core Committee and MUST be put to an Acceptance Vote.

File: bylaws/009-per-workflow.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+## Staffing Changes
+
+The Editor of an approved PER is empowered to add or remove members from the Working Group at any time, with notification to the Core Committee.
```
Comment (shadowhand) [inline bylaws/009-per-workflow.md:65]:
```suggestion
The Editor of an approved PER is empowered to add or remove members from the Working Group at any time. Any member changes to the Working Group MUST be declared to the Core Committee by a post on the mailing list with a link to a Pull Request that documents the change.
```

File: bylaws/009-per-workflow.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+## PER Working Group Fast-track
+
+A PER Working Group may be established with a single Entrance Vote and only three members (Editor and two others) in the following circumstances:
+
+1. A new Working Group is taking over an Abandoned PER.
+2. A new Working Group wishes to use an existing PSR as the initial version of a PER.  In this case, the assumption is that the subject matter is more suited to a PER than a PSR.
+3. A new Working Group wishes to take ownership of existing utility libraries already published that do not yet have clear ownership.
+
+Should a PSR be approved for which its Working Group has already produced one or more utility libraries, the PSR Working Group is automatically converted into a PER Working Group upon approval to cover those libraries, unless the Working Group explicitly states its intent to not do so.  For example, such utility libraries may be better managed by an existing PER for a related topic.
```
Comment (shadowhand) [inline bylaws/009-per-workflow.md:83]: This section is confusing. I understand that the intention is for a PSR working group to be converted to a PER working group after the approval of the PSR, but the wording is rather circuitous.

File: bylaws/009-per-workflow.md (hunk 3 above)
Comment (Crell) [reply bylaws/009-per-workflow.md:21]: Not in particular, beyond knee-jerk preference for strings that can be identifiers.  In this case it probably is better to let it be longer.

File: bylaws/007-psr-naming-conventions.md (hunk 6 above)
Comment (Crell) [reply bylaws/007-psr-naming-conventions.md:9]: Mainly because I was using PERs as a place for util libraries to live, and that already has an established `Fig` namespace.

If we split those up into 3 separate tracks, then I don't know off hand if it makes sense for PERs to have their own namespace or not.  Many (eg, coding standards) likely won't even have code artifacts.

File: bylaws/007-psr-naming-conventions.md (hunk 8 above)
Comment (Crell) [reply bylaws/007-psr-naming-conventions.md:20]: Yep, omission.  May change depending on how deep we decide to refactor this. (Per discussion on list.)

File: bylaws/009-per-workflow.md (hunk 9 above)
Comment (Crell) [reply bylaws/009-per-workflow.md:3]: Or just "as the PHP language and ecosystem develop"?

File: bylaws/009-per-workflow.md [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+## Accepted
+
+If the Acceptance Vote passes, then the proposal officially becomes an accepted PER.  It is considered the 1.0.0 release of that PER.
+
+## Evolution
+
+Unlike a PSR, a PER is expected to evolve over time.  Once the initial version of a PER is approved, the Working Group remains active but its minimum requirements drop to an Editor and two additional individuals.  By default, the Working Group at time of first approval continues in that role.  The Working Group may continue to develop the PER over time as the context of the recommendation evolves, such as newly released PHP language versions, additional input from third parties, etc.
+
+For any proposed change to the PER, the Editor must notify the Core Committee via a post to the mailing list of an Intent to Merge a given change.
```
[the remark this answers is not in the document]
Comment (Crell) [reply bylaws/009-per-workflow.md:57]: Mm, possibly.  My concern is that a whole crapload of stuff could be in a given release, and then if the CC objects you have to sort out how to pull out or change selected pieces.

"Ask forgiveness, not permission" makes saying "no" much harder, even in cases where saying no is the right thing to do.  And sometimes, it will be the right thing to do.

File: bylaws/009-per-workflow.md (hunk 12 above)
Comment (Crell) [reply bylaws/009-per-workflow.md:83]: Yeah, there's a lot of conditionals in this process.  The intent is that, for instance, should PSR-5 ever actually get finished, it would be a PSR that defines parsing rules.  Then after it passes it would mutate into a PER WG to define the tag library.  Or, PSR-18 may have produced a util library, but there's little reason to have a separate PER WG to maintain that when there's already the one left over from PSR-7/15/17 and the PSR-18 people want to just dump the PSR-18 code artifacts on the existing WG to handle (which is probably the best course of action).

If we do a deeper refactor as the list seems to be suggesting I will try to flesh this out further.

File: bylaws/009-per-workflow.md (hunk 9 above)
Comment (shadowhand) [reply bylaws/009-per-workflow.md:3]: I like that.

File: bylaws/009-per-workflow.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (shadowhand) [reply bylaws/009-per-workflow.md:57]: Well I would hope that the people most likely to have strong opinions about a project would be part of the Working Group. I am pretty sure that "notify the CC of any change" is going to be overly burdensome.

**Comment by Crell** [comment]
OK, big big update pushed.  I like this version a lot better.  See the updated description for the full list of meaningful changes.

**Review by PowerKiKi — CHANGES_REQUESTED** [review]
This introduce 47 cases of two spaces after a period (`.  ` instead of `. `), that should be corrected. Because it breaks consistency with existing content. And because it is trying to style Markdown that will be rendered as HTML in most cases anyway, so it is not really worth the effort.

**Comment by shadowhand** [comment]
I agree about the double-space thing, it actually makes my brain skip as I read the raw markdown.

File: bylaws/001-mission-and-structure.md [hunk 14]
```diff
@@ -71,16 +71,62 @@ Also, whilst not full members themselves, they have the ability to start any vot
...
+
+An individual may be the Sponsor of more than one Working Group simultaneously.
+
+Multiple other Core Committee members may also be on the working group if they wish.
+
+### Limited Working Group
+
+A Limited Working Group consists of an Editor and at least two other individuals.  No Sponsor is required.  Any individual may be the Editor or member of a Limited Working Group, with the exception that the Editor may not also be a Secretary.
+
+### Working Group management
```
Comment (shadowhand) [inline bylaws/001-mission-and-structure.md:96]:
```suggestion
### Working Group Management
```

**Comment by shadowhand** [comment]
These changes satisfy my previous concerns.

File: bylaws/001-mission-and-structure.md [hunk 15]
```diff
@@ -71,16 +71,62 @@ Also, whilst not full members themselves, they have the ability to start any vot
...
+
+The Editor of a Working Group is automatically the Maintainer of that Working Group's effort and output.
+
+The Maintainer of a given artifact is appointed by the Core Committee by Approval Vote.  Once a given artifact is created, the Maintainer may step down and name a replacement Maintainer at any time with Implicit Approval.  If necessary, a Decision vote to appoint a new Maintainer may be held following a suitable nomination period.
+
+The Secretaries will ensure that the Maintainer has the necessary access and resources to develop the artifact, such as access to a GitHub repository, mailing list, chat room, or similar such tools.
+
+## Artifacts
+
+The PHP FIG produces three primary forms of output artifact that are subject to established workflow processes.  This list is not exhaustive of the activity or publication that the PHP FIG may engage in.
```
Comment (Jean85) [inline bylaws/001-mission-and-structure.md:128]: Please use the dashed form in any place in our documents:
```suggestion
The PHP-FIG produces three primary forms of output artifact that are subject to established workflow processes.  This list is not exhaustive of the activity or publication that the PHP-FIG may engage in.
```

File: bylaws/001-mission-and-structure.md [hunk 16]
```diff
@@ -71,16 +71,62 @@ Also, whilst not full members themselves, they have the ability to start any vot
...
+
+* A PHP Standard Recommendation (PSR) defines an interoperability standard that establishes a "contract" between providers and consumers.  The goal of any PSR is to encourage or facilitate cross-project collaboration and standardization.  PSRs are developed to a particular end-state and, once approved, are considered frozen in time to provide a stable and reliable target for implementers.  PSRs are developed by a Working Group.
```
Comment (Jean85) [inline bylaws/001-mission-and-structure.md:130]: Should we reference the evolution bylaw as the only exception to this "frozen" situation?

**Review by Jean85 — CHANGES_REQUESTED** [review]

File: bylaws/001-mission-and-structure.md (hunk 16 above)
Comment (Crell) [reply bylaws/001-mission-and-structure.md:130]: Mm, probably.  I'll add.

We should probably also add a bunch of cross-links to the bylaws at this point, but that's probably a task for later.

File: bylaws/001-mission-and-structure.md (hunk 15 above)
Comment (Crell) [reply bylaws/001-mission-and-structure.md:128]: Curiously, the bylaws seem to use PHP FIG almost exclusively, not PHP-FIG.  (It's 15 to 0 in the mission-and-structure bylaw right now.)  That may be worth changing, but seems separate from the other changes here.

Comment (Jean85) [reply bylaws/001-mission-and-structure.md:128]: If that's the case, it's a task for us @php-fig/secretaries, with no urgency.

**Review by Jean85 — APPROVED** [review]

**Review by ashnazg — APPROVED** [review]

File: bylaws/001-mission-and-structure.md [hunk 17]
```diff
@@ -71,16 +71,62 @@ Also, whilst not full members themselves, they have the ability to start any vot
...
+Every Working Group has a single Editor, who is responsible for the smooth operation of the Working Group's mission and has final authority on the group's output.  Editors are appointed by the Core Committee.  Editors are responsible for managing the development of the Working Group's efforts; for representing the Working Group in discussions with the rest of the PHP FIG; for coordinating other contributors; and for working with other members of the Working Group and Core Committee as appropriate to see an effort through the appropriate process.  Any individual may be an Editor, provided they are not also a Secretary. If the Editor of a proposal is missing for more than 45 days without notice then the Core Committee may agree upon a new Editor. The Editor is the final authority on the output of a Working Group's efforts, unless otherwise noted.
```
Comment (samdark) [inline bylaws/001-mission-and-structure.md:76]: A bit of duplication here. Not that bad but repetitive.

> and has final authority on the group's output

> is the final authority on the output of a Working Group's efforts

File: bylaws/009-naming-conventions.md [hunk 18]
```diff
@@ -0,0 +1,11 @@
+Naming conventions for code released by PHP FIG
+===============================================
+
+1. Interfaces MUST be suffixed by `Interface`: e.g. `Psr\Foo\BarInterface`.
+2. Abstract classes MUST be prefixed by `Abstract`: e.g. `Psr\Foo\AbstractBar`.
+3. Traits MUST be suffixed by `Trait`: e.g. `Psr\Foo\BarTrait`.
+4. PSR-1, 4, and 12 MUST be followed.
+5. For code released as part of a PSR, the vendor namespace MUST be `Psr` and the Composer package name MUST be `psr/<package>` (e.g., `psr/log`).
+6. For code released as part of a PER or any other Auxiliary Resources, the vendor namespace MUST be `Fig` and the Composer package name MUST be `fig/<package>` (e.g., `fig/cache-util`).
```
Comment (samdark) [inline bylaws/009-naming-conventions.md:9]: Why not `Per`/`per`?

**Review by samdark — APPROVED** [review]

File: bylaws/009-naming-conventions.md (hunk 18 above)
Comment (Crell) [reply bylaws/009-naming-conventions.md:9]: Mainly because `Fig` is already established and in use, and I see little reason to add another namespace for us to manage. :smile: If we were starting over today, I think we'd probably use `fig/` as the namespace for everything, including PSRs.

File: bylaws/009-naming-conventions.md [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+7. There MUST be a package/second-level namespace in relation with the PSR or PER that  covers the code.
```
Comment (zoispag) [inline bylaws/009-naming-conventions.md:10]:
```suggestion
7. There MUST be a package/second-level namespace in relation to the PSR or PER that covers the code.
```

**Review by Jean85 — APPROVED** [review]
Vote is closed, the bylaw change is approved!

Links to the votes:
 * https://groups.google.com/g/php-fig/c/llG3bZ3Z5jo
 * https://groups.google.com/g/php-fig/c/sNHGSClcwB0

**Comment by Jean85** [comment]
@Crell feel free to port the remainder suggested changes in follow-up PRs, since they all seems to be all trivial changes.
